tests/model_fields/test_jsonfield.py

```diff
@@ -0,0 +1,676 @@
+        self._set_encoder_decoder(StrEncoder, None)
+        obj = JSONModel(value=value)
+        with transaction.atomic(), self.assertRaises(DatabaseError):
+            obj.save()
+
+
+class TestModelFormField(SimpleTestCase):
+    def test_formfield(self):
+        model_field = models.JSONField()
+        form_field = model_field.formfield()
+        self.assertIsInstance(form_field, forms.JSONField)
+
+    def test_formfield_custom_encoder_decoder(self):
+        model_field = models.JSONField(encoder=DjangoJSONEncoder, decoder=CustomDecoder)
+        form_field = model_field.formfield()
+        self.assertIs(form_field.encoder, DjangoJSONEncoder)
+        self.assertIs(form_field.decoder, CustomDecoder)
+
+
+class TestSerialization(SimpleTestCase):
+    test_data = (
+        '[{"fields": {"value": %s}, '
+        '"model": "model_fields.jsonmodel", "pk": null}]'
+    )
+    test_values = (
+        # (Python value, serialized value),
+        ({'a': 'b', 'c': None}, '{"a": "b", "c": null}'),
+        ('abc', '"abc"'),
+        ('{"a": "a"}', '"{\\"a\\": \\"a\\"}"'),
+    )
+
+    def test_dumping(self):
+        for value, serialized in self.test_values:
+            with self.subTest(value=value):
+                instance = JSONModel(value=value)
+                data = serializers.serialize('json', [instance])
+                self.assertJSONEqual(data, self.test_data % serialized)
+
+    def test_loading(self):
+        for value, serialized in self.test_values:
+            with self.subTest(value=value):
+                instance = list(
+                    serializers.deserialize('json', self.test_data % serialized)
+                )[0].object
+                self.assertEqual(instance.value, value)
+
+
+class TestSaveLoad(TestCase):
+    def test_null(self):
+        obj = NullableJSONModel(value=None)
+        obj.save()
+        obj.refresh_from_db()
+        self.assertEqual(
+            obj.value,
+            '' if connection.features.interprets_empty_strings_as_nulls else None,
+        )
+
+    @skipUnlessDBFeature('supports_primitives_in_json_field')
+    def test_json_null_different_from_sql_null(self):
+        json_null = NullableJSONModel.objects.create(value=Value('null'))
+        json_null.refresh_from_db()
+        sql_null = NullableJSONModel.objects.create(value=None)
+        sql_null.refresh_from_db()
+
+        # They are different in the database ('null' vs NULL).
+        self.assertSequenceEqual(
+            NullableJSONModel.objects.filter(value=Value('null')),
+            [json_null]
+        )
+        self.assertSequenceEqual(
+            NullableJSONModel.objects.filter(value=None),
+            [json_null]
+        )
+        self.assertSequenceEqual(
+            NullableJSONModel.objects.filter(value__isnull=True),
+            [sql_null]
+        )
+        # They are equal in Python (None).
+        self.assertEqual(json_null.value, sql_null.value)
```

Comment:
I really don't understand what we're testing here :thinking:

Reply:
We're testing that it is possible (while not recommended) to store a JSON `null` value, which is different from SQL `NULL` value. It's possible by using `Value('null')` (or by `RawSQL`) to store the JSON `null` value in the model and then save it to the database. However, when it's retrieved from the database, we call `json.loads('null')` on the value, which will return `None`. Therefore, after retrieval, it will be just like SQL `NULL` (which we don't call `json.loads()` on, just return `None`).

I didn't know whether it should be deserialized as `Value('null')` again, or if we should just ignore this case.

Reply:
IMO this behavior is extremely tricky and unexpected. I would expect that `Value('null')` is saved as a string `'null'` :thinking:

Reply:
> IMO this behavior is extremely tricky and unexpected. I would expect that `Value('null')` is saved as a string `'null'`

If what you mean by "saved" is on the database level, that's what would currently happen (it's saved as the *database string* `'null'`, but interpreted as JSON value `null` by the JSON functions). However, it's also different from the Python string `'null'` which will be encoded as `'"null"'` on the database.

We *could* change `from_db_value` so `'null'` gets deserialized as `Value('null')` (probably only if `self.decoder` isn't set) instead of calling `json.loads()`, but I'm not sure if that's a better option.

Reply:
I think you are correct: it is tricky and unexpected, but it's what the databases do. The string of `'null'` is turned into a JSON object (with value `null`). When that is loaded from the database, and turned back into a Python object, it's converted to `None`. Which, unfortunately, is what `NULL::JSONB` is also converted to.

I _guess_ we could convert it to `Value('null')` instead, but that's probably going to be even less fun.

Reply:
Well, JSON has its own `null` type. It is different from the string `'null'`. Please check this:

```
>>> json.dumps({'a': None, 'b': 'null', 'c': 52})
'{"a": null, "b": "null", "c": 52}'
>>> json.loads('{"a": null, "b": "null", "c": 52}')
{'a': None, 'b': 'null', 'c': 52}
```

If the string `null` and `None` would stringify to the same, it would be a real problem IMO.
https://www.bbc.com/future/article/20160325-the-names-that-break-computer-systems

Reply:
> Well, JSON has its own `null` type. It is different from the string `'null'`. Please check this:
> 
> ```
> >>> json.dumps({'a': None, 'b': 'null', 'c': 52})
> '{"a": null, "b": "null", "c": 52}'
> >>> json.loads('{"a": null, "b": "null", "c": 52}')
> {'a': None, 'b': 'null', 'c': 52}
> ```
> 
> If the string `null` and `None` would stringify to the same, it would be a real problem IMO.
> https://www.bbc.com/future/article/20160325-the-names-that-break-computer-systems

Actually, we're talking about the possibility of storing the `null` primitive on the field itself. So, it's not something like `{"a": null}`, but just `null`. Those cases you mention are already handled correctly (we actually also use `json.dumps` and `json.loads`).

Normally, Django stores Python `None` as SQL `NULL`. However, since we can store JSON primitives on `JSONField`s, we're wondering how we should handle the `null` primitive. Hope that clears it up @mbeijen!